modules/standard/IO.chpl

```diff
@@ -8086,67 +8108,231 @@ proc fileReader.readBytes(maxSize: int): bytes throws {
+  var guessReadSize:c_ssize_t = 0;
+  var fp: qio_file_ptr_t = nil;
+  qio_channel_get_file_ptr(ch._channel_internal, fp);
+  var fileLen:int(64) = -1;
+  if fp {
+    fileLen = qio_file_length_guess(fp);
+  }
+  if pos >= 0 && fileLen >= 1 && fileLen > pos {
+    guessReadSize = (fileLen - pos):c_ssize_t;
+  }
+
+  // limit the size to read by maxBytes
+  if guessReadSize > maxChars {
+    guessReadSize = maxChars;
+  }
+
+  assert(guessReadSize >= 0);
+
+  return guessReadSize;
+}
+
+// note: len is in codepoints when t == string, and in bytes otherwise
+private proc readBytesImpl(ch: fileReader, ref out_var: bytes, len: int(64)) : (errorCode, int(64))
   throws
 {
-
   var err:errorCode = 0;
   var lenread:int(64);
 
   on ch._home {
-    var tx:c_ptrConst(c_char);
-    var lentmp:int(64);
-    var actlen:int(64);
-    var uselen:c_ssize_t;
+    try ch.lock(); defer { ch.unlock(); }
 
-    if len == -1 then uselen = max(c_ssize_t);
-    else {
-      uselen = len:c_ssize_t;
-      if c_ssize_t != int(64) then assert( len == uselen );
+    // note the current channel position
+    var pos = qio_channel_offset_unlocked(ch._channel_internal);
+
+    // Compute the maximum amount we could read as a 'c_ssize_t'
+    // based upon 'len' and the channel's region.
+    // This handles len==-1 as well as a channel with a bounded region.
+    const maxBytes:c_ssize_t = computeMaxBytesToRead(ch, len, pos, bytes);
+
+    // Compute a guess as to the size to read based on the file length.
+    // This is only a guess & it's possible it will be out of date
+    // by the time we actually read.
+    // We'll use this guess to decide how much to allocate up-front.
+    // We don't want to allocate all of 'len' up front if it's bigger
+    // than the observed channel size or the initial file size.
+    const guessReadSize:c_ssize_t =
+      computeGuessReadSize(ch, maxBytes, pos)+1; // +1 for trailing \0
+
+    // proactively allocate 'guessReadSize'
+    var buff: bufferType = nil;
+    var buffSz = 0;
+    var n:c_ssize_t = 0; // how many bytes have we read into buff?
+    (buff, buffSz) = bufferAlloc(guessReadSize); // room for trailing \0
+
+    // then try to read repeatedly until we have read 'maxBytes' or reach EOF
+    while n < maxBytes {
+      var locErr:errorCode = 0;
+      var amtRead:c_ssize_t = 0;
+      if n >= buffSz {
+        // if we need more room in the buffer, grow it
+        // this will happen if we have not read all of 'maxBytes' yet
+        // but there is more data in the file (as when guessReadSize
+        // was innacurate for one reason or another)
+        var requestSz = 2*buffSz;
+        // make sure to at least request 16 bytes
+        if requestSz < n + 16 then requestSz = n + 16;
+        // but don't ever ask for more bytes than maxBytes + 1
+        if maxBytes < max(c_ssize_t) && requestSz > maxBytes + 1 then
+           requestSz = maxBytes + 1;
```

Comment:
shouldn't this be: 
```chapel
if maxBytes < max(c_ssize_t) && requestSz + n > maxBytes + 1 then
   requestSz = maxBytes + 1 - n;
```

so we don't allocate more buffer space than is needed to store `len` total bytes.

Reply:
Here `requestSz` is going to be the new size of the buffer which is `realloc`'d. So we can't subtract `n` here because that would cause us to lose the data read in earlier iterations of the loop.

`maxBytes` is computed from `len`, so we should not be allocating more than `len+1` (assuming `len` is positive) here.